File: 2024/day-01.rb
#!/usr/bin/env ruby

# https://adventofcode.com/2024/day/1

require "bundler/setup"
Bundler.require

first, second = File
  .foreach("#{__dir__}/day-01.txt")
  .map { |x| x.split.map(&:to_i) }
  .transpose

first.sort!
second.sort!

result = first.zip(second)
  .reduce(0) { |acc, (a, b)| acc + (a - b).abs }

puts result

File: 2024/day-01.exs
#!/usr/bin/env elixir

# https://adventofcode.com/2024/day/1

{first, second} =
  File.stream!("day-01.txt")
  |> Enum.map(fn line ->
    [a, b] = String.split(line)
    {String.to_integer(a), String.to_integer(b)}
  end)
  |> Enum.unzip()

part1 =
  Enum.zip(Enum.sort(first), Enum.sort(second))
  |> Enum.reduce(0, fn {a, b}, acc ->
    acc + abs(a - b)
  end)

IO.inspect(part1)

second_freq = Enum.frequencies(second)

part2 =
  Enum.frequencies(first)
  |> Enum.reduce(0, fn {k, v}, acc ->
    acc + k * v * Map.get(second_freq, k, 0)
  end)

IO.inspect(part2)
